modules/server/src/main/scala/scaladex/server/service/SonatypeService.scala

```diff
@@ -61,26 +78,21 @@ class SonatypeService(
         .mapSync(id => findAndIndexMissingArtifacts(groupId, id, knownRefs))
     } yield result.sum
 
-  private def findAndIndexMissingArtifacts(
-      groupId: GroupId,
-      artifactId: ArtifactId,
-      knownRefs: Set[MavenReference]
-  ): Future[Int] =
+  def syncOne(groupId: GroupId, artifactNameOpt: Option[Artifact.Name]): Future[String] =
     for {
-      versions <- sonatypeService.getAllVersions(groupId, artifactId)
-      mavenReferences = versions.map(v =>
-        MavenReference(groupId = groupId.value, artifactId = artifactId.value, version = v.toString)
-      )
-      missingVersions = mavenReferences.filterNot(knownRefs)
-      _ = if (missingVersions.nonEmpty)
-        logger.warn(s"${missingVersions.size} artifacts are missing for ${groupId.value}:${artifactId.value}")
-      missingPomFiles <- missingVersions.map(ref => sonatypeService.getPomFile(ref).map(_.map(ref -> _))).sequence
-      publishResult <- missingPomFiles.flatten.mapSync {
+      mavenReferenceFromDatabase <- database.getAllMavenReferences()
+      result <- findAndIndexMissingArtifacts(groupId, artifactNameOpt, mavenReferenceFromDatabase.toSet)
+    } yield s"Inserted $result poms"
+
+  def updateAllArtifacts(): Future[String] =
+    for {
+      mavenReferences <- database.getAllMavenReferences()
+      _ = logger.info(s"${mavenReferences.size} artifacts will be synced for new metadata.")
+      poms <- mavenReferences.map(ref => sonatypeService.getPomFile(ref).map(_.map(ref -> _))).sequence
```

Comment:
As you pointed out, this is going to OOM. You cannot retain all pom files in memory, instead we need to download one pom file at a time, and to publish it before we download another one. The GC should be able to remove the pom file from the heap as soon as it has been published.

I propose something like this:
```scala
def updateArtifact(ref: MavenReference): Future[Boolean] =
  // download the pom file and publish the artifact
  // return true if success or false if pom file is missing
```

And then call
```
  statuses <- mavenReferences.mapSync(updateArtifact)
```

`mapSync` is defined in `ScalaExtensions`, it is like `.map(???).sequence`, except it does one element at a time, instead of all in parallel.

Reply:
update the PR with mapSync